### PR TITLE
Replace toy task with fashion MNIST

### PR DESCRIPTION
**Pull request by jerpint**
Currently the toy task is to sum numbers. While it runs fast, it is not ideal for visualizing results, testing confusion matrices, displaying images, etc.

We are now replacing this with a more "visual" task, e.g. fashionMNIST

### CODE REVIEW

File: {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/data/data_loader.py [hunk 1]
```diff
@@ -1,75 +1,59 @@
...
+        transform: typing.List = None,
     ):
-        """Initialize MyDataset.
+        """Initialize Dataset.
 
         Args:
-            input_data (np.array): Input data.
-            target_data (np.array): Target data.
+            images (np.array): Image data [batch, height, width].
+            labels (np.array): Target data [batch,].
```
Comment (hbertrand) [inline {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/data/data_loader.py:28]: General question about docstrings: what's the point of putting them here when we already put them in typehints? It feels redundant.

File: {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/data/data_loader.py [hunk 2]
```diff
@@ -79,47 +63,38 @@ def __init__(
...
+
+        if stage == "fit" or stage is None:
+            self.train_dataset = FashionMnistDS(
+                raw_data.train_images, raw_data.train_labels, transform
+            )
+            self.val_dataset = FashionMnistDS(
+                raw_data.val_images, raw_data.val_labels, transform
+            )
+        if stage == "test" or stage is None:
+            self.val_dataset = FashionMnistDS(
```
Comment (hbertrand) [inline {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/data/data_loader.py:86]: should be `test_dataset`

File: {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/data/data_loader.py [hunk 3]
```diff
@@ -79,47 +63,38 @@ def __init__(
...
+                raw_data.test_images, raw_data.test_labels, transform
+            )
 
     def train_dataloader(self) -> DataLoader:
         """Creates the training dataloader using the training data parser."""
-        return DataLoader(self.train_data_parser, batch_size=self.batch_size, shuffle=True)
+        return DataLoader(self.train_dataset, batch_size=self.batch_size, shuffle=True)
```
Comment (hbertrand) [inline {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/data/data_loader.py:92]: For all dataloader, should we have a num_workers argument by default?

File: {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/data/data_preprocess.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+import numpy as np
+
+logger = logging.getLogger(__name__)
+
+
+BASE_URL = "https://github.com/zalandoresearch/fashion-mnist/raw/master/data/fashion/"
+
+
+class FashionMnistParser:
```
Comment (hbertrand) [inline {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/data/data_preprocess.py:15]: I'm guessing this is a standard thing you took from somewhere? I don't need to check how you parse the gzip?

File: {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/models/model_loader.py [hunk 5]
```diff
@@ -16,8 +16,10 @@ def load_model(hyper_params):  # pragma: no cover
     """
     architecture = hyper_params['architecture']
     # __TODO__ fix architecture list
-    if architecture == 'my_model':
-        model_class = MyModel
+    if architecture == 'simple_mlp':
+        model_class = SimpleMLP
+    elif architecture == 'simple_cnn':
+        model_class = SimpleCNN
```
Comment (hbertrand) [inline {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/models/model_loader.py:22]: This is where hydra would be useful...

Comment (mirkobronzimila) [inline {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/models/model_loader.py:22]: I would keep only one model here - the task will change in any case, and we will pretty much never use the models in this toy experiment..so, better having to modify only one (less work).
It goes back to the philosophy of not over-engineering the project template.
What do you think?

Comment (jerpint) [reply {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/models/model_loader.py:22]: yes i originally kept it to be sure performance increased but I think removing it is the right call, otherwise we could perhaps use it as an Orion parameter for HP search?

File: {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/data/data_loader.py (hunk 1 above)
Comment (jerpint) [reply {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/data/data_loader.py:28]: I agree it is redundant, but follows the google docstrings style, I wonder if there is a PEP about this?

File: {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/data/data_loader.py (hunk 2 above)
Comment (jerpint) [reply {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/data/data_loader.py]: nice catch

File: {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/data/data_preprocess.py (hunk 4 above)
Comment (jerpint) [reply {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/data/data_preprocess.py:15]: yes I linked to the stack overflow post in the docstrings, See: https://stackoverflow.com/questions/40427435/extract-images-from-idx3-ubyte-file-or-gzip-via-python # noqa

File: {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/data/data_loader.py (hunk 3 above)
Comment (jerpint) [reply {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/data/data_loader.py]: added a `num_workers` argument to the config file

File: {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/models/model_loader.py (hunk 5 above)
Comment (mirkobronzimila) [reply {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/models/model_loader.py:22]: good point, but we can use `learning_rate` or others as Orion example..
This one is better to remove because of the associated additional code (so we do not have to mantain it, and also we have less stuff to delete once we instantiate the project).

File: {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/data/data_loader.py (hunk 1 above)
Comment (mirkobronzimila) [reply {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/data/data_loader.py:29]: Cannot find anything in google, but it's a good point to use the type in the typehint, so that we have an "easy" switch to pytype.
So, if we have no disadvantages, Hadrien's point seems the best approach.

File: {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/models/model_loader.py (hunk 5 above)
Comment (jerpint) [reply {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/models/model_loader.py]: I removed it for now

File: {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/data/data_loader.py (hunk 1 above)
Comment (jerpint) [reply {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/data/data_loader.py:29]: they are already as a type for pytype, just added in the docstring, i use an autodocstring generator that uses google formatting and it auto infers the type from the typehint

**Review by hbertrand — APPROVED** [review]

File: {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/main.py [hunk 6]
```diff
@@ -9,7 +9,7 @@
 import yaml
 from yaml import load
 
-from {{cookiecutter.project_slug}}.data.data_loader import load_data
+from {{cookiecutter.project_slug}}.data.data_loader import FashionMnistDM
```
Comment (marondeau-mila) [inline {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/main.py:12]: Maybe we should rename `FashionMnistDM` to `DataModuleFactory` and add a `__init__.py` to `data/`, with
```
from . data_loader import FashionMnistDM
__all__ = [FashionMnistDM]
```
That way, it would be a generic name we don't need to change, the name says what it does, and we can type less.

File: {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/main.py [hunk 7]
```diff
@@ -119,7 +119,7 @@ def run(args, data_dir, output_dir, hyper_params):
 
     log_exp_details(os.path.realpath(__file__), args)
 
-    datamodule = load_data(data_dir, hyper_params)
+    datamodule = FashionMnistDM(data_dir, hyper_params)
     model = load_model(hyper_params)
```
Comment (marondeau-mila) [inline {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/main.py:123]: We should probably change the name of `model` to something that indicate it's a LightingModule, or at least indicates it's not simply a torch.nn.Module; I like calling the LightingModule a `Task` (e.g., `ClassifyImage` or `SegmentImage`.)
`load_model` should be updated to match `load_data`, with the same "put factory in the name" comment.

File: {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/models/model_loader.py [hunk 8]
```diff
@@ -16,8 +16,8 @@ def load_model(hyper_params):  # pragma: no cover
     """
     architecture = hyper_params['architecture']
     # __TODO__ fix architecture list
```
Comment (marondeau-mila) [inline {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/models/model_loader.py:18]: This whole thing should be a factory class, with the HP used to initialize. We can then easily change the class to use Hydra, Omniconf, etc., depending on the project, and easily add model classes as needed.

File: {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/models/my_model.py [hunk 9]
```diff
@@ -1,7 +1,7 @@
 import logging
 import typing
 
-import torch
+from torch import nn
 import pytorch_lightning as pl
 
 from {{cookiecutter.project_slug}}.models.optim import load_loss, load_optimizer
```
Comment (marondeau-mila) [inline {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/models/my_model.py:7]: `load_loss` and `load_optimizer` should be factories.

File: {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/models/my_model.py [hunk 10]
```diff
@@ -58,36 +58,38 @@ def test_step(self, batch, batch_idx):
...
 
 
-class MyModel(BaseModel):  # pragma: no cover
+class SimpleMLP(BaseModel):  # pragma: no cover
     """Simple Model Class.
 
     Inherits from the given framework's model class. This is a simple MLP model.
     """
-
     def __init__(self, hyper_params: typing.Dict[typing.AnyStr, typing.Any]):
```
Comment (marondeau-mila) [inline {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/models/my_model.py:66]: I think it's better to have a `Task`, which combines a `Loss`, a `Model`, and an `Optimizer`, all of which are initialized by factories.
```
def __init__(self, model_factory, loss_factory, opt_factory):
    super().__init__()  # Needed because this is still a Module
    self.save_hyperparameters()  # This will automatically save the factories, which should be reloaded automagically later.
```

File: {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/models/optim.py [hunk 11]
```diff
@@ -17,14 +17,14 @@ def load_optimizer(hyper_params, model):  # pragma: no cover
     Returns:
         optimizer (obj): The optimizer for the given model
     """
-    optimizer_name = hyper_params['optimizer']
```
Comment (marondeau-mila) [inline {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/models/optim.py:20]: Usual comment RE:factories.

**Review by marondeau-mila — APPROVED** [review]

File: {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/models/model_loader.py (hunk 5 above)
Comment (mirkobronzi) [reply {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/models/model_loader.py]: thanks

**Comment by marondeau-mila** [comment]
They are a few changes I believe we should make (dataclass, factories, names), but they should be kept for future PR. This one is large enough, and more focussed discussion is needed.

File: {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/data/data_loader.py [hunk 12]
```diff
@@ -79,47 +64,55 @@ def __init__(
...
-    Args:
-        data_dir (str): path to the folder containing the data
-        hyper_params (dict): hyper parameters from the config file
-
-    Returns:
-        datamodule (obj): the data module used to prepare/instantiate data loaders.
-    """
-    # __TODO__ if you have different data modules, add whatever code is needed to select them here
-    return MyDataModule(data_dir, hyper_params)
+        return DataLoader(
```
[the remark this answers is not in the document]
Comment (jerpint) [reply {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/data/data_loader.py:113]: `drop_last=False` is the default value: https://pytorch.org/docs/stable/data.html

[the remark this answers is not in the document]
Comment (marondeau-mila) [reply {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/data/data_loader.py:113]: I prefer to have that one explicit. Save a check when debugging.

File: {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/data/data_loader.py [hunk 13]
```diff
@@ -1,75 +1,60 @@
...
-            index (int): Get index item from the dataset.
+            idx (int): Get index item from the dataset.
         """
-        target_example = self.target_data[index]
-        input_example = self.input_data[index]
-        return input_example, target_example
+        image = self.images[idx]
 
+        if self.transform:
+            image = self.transform(image)
```
[the remark this answers is not in the document]
Comment (jerpint) [reply {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/data/data_loader.py:55]: fixed 👍

File: {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/data/data_loader.py [hunk 14]
```diff
@@ -79,47 +64,55 @@ def __init__(
     ):
         """Validates the hyperparameter config dictionary and sets up internal attributes."""
         super().__init__()
+        check_and_log_hp(["batch_size", "num_workers"], hyper_params)
         self.data_dir = data_dir
-        self.batch_size = hyper_params['batch_size']
-        self.train_data_parser, self.dev_data_parser = None, None
-
-    def prepare_data(self):
```
[the remark this answers is not in the document]
Comment (jerpint) [reply {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/data/data_loader.py:86]: not sure what you mean here, pytorch lightning takes care of calling `setup()` i think, this method was deprecated

File: {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/main.py (hunk 7 above)
Comment (jerpint) [reply {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/main.py:123]: i agree with you on this, but this is sadly the `pytorch_lightning` way of doing things, see https://pytorch-lightning.readthedocs.io/en/stable/common/lightning_module.html

File: {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/data/data_loader.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (marondeau-mila) [reply {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/data/data_loader.py:86]: I mean, in `main.py`, we have code that copy files from scratch. This should be in `prepare_data`, which, unlike setup, is called exactly once during multi-gpu training. That's where anything that manipulate the files (download, extracts, etc.) should be done.

File: {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/main.py (hunk 7 above)
Comment (marondeau-mila) [reply {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/main.py:123]: You can do both: the Lightning module contains the model, but does not know the details. Easy with a factory. All the module does is implement `forward` as:
```
def forward(self, image: FloatTensor) -> FloatTensor:
  return self.model(image)
```
Then, we can change the model by changing the factory we used to initialize the module:
```
task = Task(..., model_factory=resnet_factory)
```
It's more for future work, other than the names.
